### PR TITLE
Lem -> Coq conversion

### PR DESCRIPTION
**Pull request by mrsmkl**
evm.lem became uglier, need to fix bugs lem to improve.
For recursive definitions that Coq didn't accept, just added a new definition for Coq, should be quite easy to modify Isabelle proofs to work with new version.

Includes a termination proof and new lem files word32.lem and word64.lem.

Works with Coq version 8.6, need to modify makefiles in lem to make lem be able to compile coq-lib.

### CODE REVIEW

File: lem/keccak.lem [hunk 1]
```diff
@@ -124,18 +128,21 @@ let keccakf_rounds : nat = 24
...
     let quo = num / checked in
     let taken = quo * checked in
-    (quo <> 0) :: boolList64_inner (num - taken) (rest - 1)
+    (quo <> 0) :: boolList64_inner (num - taken) rest
+end
 
 val boolListFromWord64 : word64 -> list bool
 let boolListFromWord64 w = boolList64_inner w 64
 declare isabelle target_rep function boolListFromWord64 = `to_bl`
+*)
```
Comment (pirapira) [inline lem/keccak.lem]: You can remove these parts if they are not used.

File: lem/evm.lem [hunk 2]
```diff
@@ -622,17 +691,20 @@ let Gmid = 8
 val Ghigh : integer
 let Ghigh = 10
 
+val magic_block : natural
```
Comment (pirapira) [inline lem/evm.lem]: Maybe call it `eip150_block`?

File: lem/evm.lem [hunk 3]
```diff
@@ -2030,7 +2075,7 @@ let empty_memory = (fun _ -> 0)
 
 indreln [build_vctx_called : account_state -> call_env -> variable_ctx -> bool]
   vctx_called : forall bal a env origin ext block gas existence.
-    word256ToNatural block.block_number >= 2463000 && (* only the newest version of EVM matters *)
+    word256ToNatural block.block_number >= 2463*1000 && (* only the newest version of EVM matters *)
```
Comment (pirapira) [inline lem/evm.lem]: You can use the new name here as well.

File: tester/lem [hunk 4]
```diff
@@ -1 +0,0 @@
-../lem/
```
Comment (pirapira) [inline tester/lem]: Is this intentional?

**Comment by pirapira** [comment]
@mrsmkl will you fix the conflicting files, or shall I?

File: tester/lem (hunk 4 above)
Comment (mrsmkl) [inline tester/lem]: I had to remove it because ocamlbuild went to infinite loop (win10 linux subsystem).

Comment (pirapira) [inline tester/lem]: I didn't know that happens.  I need to think other ways (maybe `make` does a copy or something).

Comment (pirapira) [inline tester/lem]: Created #332 about this.

**Comment by pirapira** [comment]
Great work!  Thanks a lot!  I think we can mark v0.0.2.

**Pull request merged**